File: cpp/73_counting_fractions_in_a_range.cpp
#include <iostream>

int counting_fractions_in_a_range(int a, int b, int c, int d){
    // a/b and c/d are adjacent fractions in the set of reduced proper fractions with denominator <= 12000
    if (b + d > 12'000) return 0;
    // otherwise, mediant (a + c)/(b + d) exists between a/b and c/d with denominator <= 12000
    // count fractions between a/b and (a + c)/(b + d)
    // and fractions between (a + c)/(b + d) and c/d
    return 1 + counting_fractions_in_a_range(a, b, a + c, b + d) + counting_fractions_in_a_range(a + c, b + d, c, d);
}

int main(){
    std::cout << counting_fractions_in_a_range(1, 3, 1, 2) << "\n";
}

File: java/_73_CountingFractionsInARange.java
import java.util.*;

public class _73_CountingFractionsInARange{
    
    public static long countFractionsBetween(int a, int b, int c, int d, int n){
        if (b + d > n){
            // If the next mediant fraction between the given fractions a/b and b/d 
            // does not appear in F_n, then they are neighboring fractions in F_n.
            return 0;
        }
        
        // If b + d <= n, there exists at least one fraction between a/b and c/d in F_n.
        // There may also be fractions between the mediant and both bounds.
        return 1 + countFractionsBetween(a, b, a+c, b+d, n) + countFractionsBetween(a+c, b+d, b, d, n);
    }
    
    public static void main(String[] args){
        
        int a = 1, b = 3, c = 1, d = 2, n = 12000;
        
        long count = countFractionsBetween(a, b, c, d, n);
        
        System.out.println("Number of fractions between 1/3 and 1/2 in the set of reduced proper fractions up to d <= 12000: " + count); // 7295372
    }
}